integration/util/registry.go

```diff
@@ -0,0 +1,244 @@
+func WriteConfigFile(path string, cfg interface{}) error {
+	jsonContents, err := json.MarshalIndent(cfg, "", "\t")
+	if err != nil {
+		return fmt.Errorf("unable to marshal to json: %v", err)
+	}
+
+	// Preserve the file mode if the file already exists
+	mode, err := getFileModeOrDefault(path, configDefaultMode)
+	if err != nil {
+		return fmt.Errorf("unable to get file mode %s: %v", path, err)
+	}
+	if err = os.WriteFile(path, jsonContents, mode); err != nil {
+		return fmt.Errorf("unable to save file %s: %v", path, err)
+	}
+	return nil
+}
+
+func getFileModeOrDefault(path string, defaultMode os.FileMode) (os.FileMode, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return defaultMode, err
+	}
+	return fileInfo.Mode(), nil
+}
+
+// DeleteResources deletes all given resources and all related devices.
+func DeleteResources(cfg *TestConfiguration, resources []*Resource, deviceId, url, user, pass string) []error {
+	deleteErrors := deleteRelatedDevices(cfg, deviceId, url, user, pass)
+	// Delete in reverse order of creation
+	for i := len(resources) - 1; i >= 0; i-- {
+		r := resources[i]
+
+		if !r.delete {
+			continue
+		}
+
+		if _, err := SendDeviceRegistryRequest(nil, http.MethodDelete, r.url, r.user, r.pass); err != nil {
+			deleteErrors = append(deleteErrors, err)
+		}
+	}
+	return deleteErrors
+}
+
+func deleteRelatedDevices(cfg *TestConfiguration, viaDeviceID, url, user, pass string) []error {
+	var deleteErrors []error
+	devicesVia, err := findDeviceRegistryDevicesVia(viaDeviceID, url, user, pass)
+	if err != nil {
+		deleteErrors = append(deleteErrors, err)
+	}
+	// Digital Twin API things are created after Device Registry devices, so delete them first
+	if delErrors := deleteDigitalTwinThings(cfg, devicesVia); delErrors != nil {
+		deleteErrors = append(deleteErrors, delErrors...)
+	}
+	// Then delete Device Registry devices
+	if delErrors := deleteRegistryDevices(devicesVia, url, user, pass); delErrors != nil {
+		deleteErrors = append(deleteErrors, delErrors...)
+	}
+	return deleteErrors
+}
+
+func findDeviceRegistryDevicesVia(viaDeviceID, url, user, pass string) ([]string, error) {
+	var devicesVia []string
+
+	type registryDevice struct {
+		ID  string   `json:"id"`
+		Via []string `json:"via"`
+	}
+
+	type registryDevices struct {
+		Devices []*registryDevice `json:"result"`
+	}
+
+	contains := func(where []string, what string) bool {
+		for _, item := range where {
+			if item == what {
+				return true
+			}
+		}
+		return false
+	}
+
+	devicesJSON, err := SendDeviceRegistryRequest(nil, http.MethodGet, url, user, pass)
+	if err != nil {
+		return devicesVia, err
+	} else {
+		devices := &registryDevices{}
+		err = json.Unmarshal(devicesJSON, devices)
+		if err != nil {
+			return devicesVia, err
+		}
+		for _, device := range devices.Devices {
+			if contains(device.Via, viaDeviceID) {
+				devicesVia = append(devicesVia, device.ID)
+			}
+		}
+	}
```

Comment:
I think we can simplify this a bit:

```suggestion
	type registryDevice struct {
		ID  string   `json:"id"`
		Via []string `json:"via"`
	}

	type registryDevices struct {
		Devices []*registryDevice `json:"result"`
	}

	devicesJSON, err := SendDeviceRegistryRequest(nil, http.MethodGet, url, user, pass)
	if err != nil {
		return nil, err
	}
	devices := &registryDevices{}
	err = json.Unmarshal(devicesJSON, devices)
	if err != nil {
		return nil, err
	}

	var devicesVia []string
	for _, device := range devices.Devices {
		for _, via := range device.Via {
			if via == viaDeviceID {
				devicesVia = append(devicesVia, device.ID)
				break
			}
		}
	}
```